sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageStreamSource.java

```diff
@@ -215,23 +223,101 @@ public T getCurrent() throws NoSuchElementException {
+              // use setFraction().
+              .setUnknownFields(
+                  UnknownFieldSet.newBuilder()
+                      .addField(
+                          2,
+                          UnknownFieldSet.Field.newBuilder()
+                              .addFixed32(java.lang.Float.floatToIntBits((float) fraction))
+                              .build())
+                      .build())
+              .build();
+      SplitReadStreamResponse splitResponse = storageClient.splitReadStream(splitRequest);
+
+      if (!splitResponse.hasPrimaryStream() || !splitResponse.hasRemainderStream()) {
+        // No more splits are possible!
+        Metrics.counter(
+                BigQueryStorageStreamReader.class,
+                "split-at-fraction-calls-failed-due-to-impossible-split-point")
+            .inc();
+        LOGGER.info("Stream '{}' cannot be split at {}.", source.stream.getName(), fraction);
+        return null;
+      }
+
+      // We may be able to split this source. Before continuing, we pause the reader thread and
+      // replace its current source with the primary stream iff the reader has not moved past
+      // the split point.
+      synchronized (this) {
+        BigQueryServerStream<ReadRowsResponse> newResponseStream;
+        Iterator<ReadRowsResponse> newResponseIterator;
+        try {
+          newResponseStream =
+              storageClient.readRows(
+                  ReadRowsRequest.newBuilder()
+                      .setReadPosition(
+                          StreamPosition.newBuilder()
+                              .setStream(splitResponse.getPrimaryStream())
+                              .setOffset(currentOffset + 1))
+                      .build());
+          newResponseIterator = newResponseStream.iterator();
+          newResponseIterator.hasNext();
+        } catch (FailedPreconditionException e) {
+          // The current source has already moved past the split point, so this split attempt
+          // is unsuccessful.
+          Metrics.counter(
+                  BigQueryStorageStreamReader.class,
+                  "split-at-fraction-calls-failed-due-to-bad-split-point")
+              .inc();
+          LOGGER.info(
+              "Split of stream '{}' abandoned because the primary stream is to the left of "
+                  + "the split fraction {}.",
+              source.stream.getName(),
+              fraction);
+          return null;
+        } catch (Exception e) {
+          Metrics.counter(
+                  BigQueryStorageStreamReader.class,
+                  "split-at-fraction-calls-failed-due-to-other-reasons")
+              .inc();
+          throw e;
+        }
+
+        // Cancels the parent stream before replacing it with the primary stream.
+        responseStream.cancel();
+
+        currentOffset++;
+        source = source.fromExisting(splitResponse.getPrimaryStream());
```

Comment:
Do we have to call setOffset(currentOffset + 1) again or was the call within the "try" already performed the seek ?

Reply:
The offset is not a property of the source or stream resource. It's a property of the ReadRows request/streaming RPC. Above we create a new client-server stream that begins with the right offset (responseStream) and we replace the current one with it.

However, we also save a reference to the BigQueryStorageStreamSource pointing to the primary stream, and a client can get access to that through BigQueryStorageStreamReader#getCurrentSource. Does that source need to start at whatever offset the parent reader was at when split happened OR does it have to start at 0 (i.e., retaining the property of starting where the parent stream initially started)?